src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol

```diff
@@ -0,0 +1,254 @@
+        // Mint tokens to address
+        _mint(_receiver, mintAmount);
+        // Emit event
+        emit TokensBought(_receiver, _depositAmount, mintAmount, _msgSender());
+    }
+
+    /// @dev Opens the buy functionality by setting the state variable `buyIsOpen` to true.
+    function _openBuy() internal {
+        if (buyIsOpen == true) {
+            revert BondingCurveFundingManager__BuyingAlreadyOpen();
+        }
+        buyIsOpen = true;
+        emit BuyingEnabled();
+    }
+
+    /// @dev Closes the buy functionality by setting the state variable `buyIsOpen` to false.
+    function _closeBuy() internal {
+        if (buyIsOpen == false) {
+            revert BondingCurveFundingManager__BuyingAlreadyClosed();
+        }
+        buyIsOpen = false;
+        emit BuyingDisabled();
+    }
+
+    /// @dev Sets the buy transaction fee, expressed in BPS.
+    /// @param _fee The fee percentage to set for buy transactions.
+    function _setBuyFee(uint _fee) internal {
+        if (_fee >= BPS) {
+            revert BondingCurveFundingManager__InvalidFeePercentage();
+        }
+        emit BuyFeeUpdated(_fee, buyFee);
+        buyFee = _fee;
+    }
+
+    /// @dev Calculates the deposit amount after deducting the fee.
+    /// The function takes a deposit amount and a fee percentage to calculate
+    /// the net deposit amount after fee deduction.
+    /// @param _depositAmount The original amount to be deposited.
+    /// @param _feePct The fee percentage to be deducted, represented in basis points.
+    /// @return depositAmountMinusFee The deposit amount after fee has been deducted.
+    function _calculateFeeDeductedDepositAmount(
+        uint _depositAmount,
+        uint _feePct
+    ) internal pure returns (uint depositAmountMinusFee) {
+        if (_feePct > BPS) {
+            revert BondingCurveFundingManager__InvalidFeePercentage();
+        }
+        // Calculate fee amount
+        uint feeAmount = (_depositAmount * _feePct) / BPS;
+        // Subtract fee amount from deposit amount
+        depositAmountMinusFee = _depositAmount - feeAmount;
+    }
+
+    /// @dev Issues tokens based on the deposit amount.
+    /// This function utilizes another internal function, `_issueTokensFormulaWrapper`,
+    /// to determine how many tokens should be minted.
+    /// @param _depositAmount The amount of funds deposited for which tokens are to be issued.
+    /// @return mintAmount The number of tokens that will be minted.
+    function _issueTokens(uint _depositAmount)
+        internal
+        returns (uint mintAmount)
+    {
+        mintAmount = _issueTokensFormulaWrapper(_depositAmount);
+    }
+
+    /// @dev Sets the number of decimals for the token.
+    /// This function updates the `tokenDecimals` state variable.
+    /// @param _decimals The number of decimals to set for the token.
+    function _setTokenDecimals(uint8 _decimals) internal {
+        if (_decimals == 0) {
```

Comment:
Decision moving forwards is influenced by answers from the following comment:
- https://github.com/InverterNetwork/inverter-contracts/pull/350#discussion_r1360626522

Reply:
The token decimal constant has been removed and the value can now be set again during init, which addresses your second comment. In regard to your first comment, we decided to remove the input validation from the base contract because as you point out, there might be a future implementation where a `decimal == 0` is needed. However to keep the flexibility for the implementation contract to enforce a specific issuance token decimal, we made the internal function `_setTokenDecimals` virtual.

It now get overridden within the `BancorVirtualSupplyBondingCurveFundingManager` to add a check for `decimals > 7` to ensure that there is no destructive precision loss when using `PPM` for calculation within the Bancor Formula contract